File: unii-app/lib/data/models/activity.dart
class LngLat {
  const LngLat({required this.lng, required this.lat});

  factory LngLat.fromJson(Map<String, dynamic> json) => LngLat(
    lng: (json['lng'] as num).toDouble(),
    lat: (json['lat'] as num).toDouble(),
  );

  final double lng;
  final double lat;

  Map<String, dynamic> toJson() => {'lng': lng, 'lat': lat};
}

class Activity {
  const Activity({
    required this.id,
    required this.teamId,
    required this.creatorId,
    required this.title,
    required this.location,
    required this.visibility,
    this.locationName,
    this.startTime,
    this.endTime,
    this.content,
    this.notice,
    this.createdAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json) => Activity(
    id: (json['id'] as num).toInt(),
    teamId: (json['team_id'] as num).toInt(),
    creatorId: (json['creator_id'] as num).toInt(),
    title: json['title'] as String,
    location: LngLat.fromJson(json['location'] as Map<String, dynamic>),
    visibility: json['visibility'] as String,
    locationName: json['location_name'] as String?,
    startTime: json['start_time'] as String?,
    endTime: json['end_time'] as String?,
    content: json['content'] as String?,
    notice: json['notice'] as String?,
    createdAt: json['created_at'] as String?,
  );

  final int id;
  final int teamId;
  final int creatorId;
  final String title;
  final LngLat location;
  final String? locationName;
  final String? startTime;
  final String? endTime;
  final String? content;
  final String? notice;
  final String visibility;
  final String? createdAt;

  bool get isPublic => visibility == 'public';
}

File: unii-server/src/dto/activity.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::model::activity::ActivityRow;

/// Coordinate pair — WGS84, `(lng, lat)` order matches plan.md §6.1.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
pub struct LngLat {
    pub lng: f64,
    pub lat: f64,
}

impl LngLat {
    pub fn is_valid(&self) -> bool {
        (-180.0..=180.0).contains(&self.lng) && (-90.0..=90.0).contains(&self.lat)
    }
}

#[derive(Debug, Deserialize)]
pub struct CreateActivityReq {
    pub title: String,
    pub location: LngLat,
    pub location_name: Option<String>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub content: Option<String>,
    pub notice: Option<String>,
    pub visibility: String,
}

#[derive(Debug, Default, Deserialize)]
pub struct UpdateActivityReq {
    pub title: Option<String>,
    pub location: Option<LngLat>,
    pub location_name: Option<String>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub content: Option<String>,
    pub notice: Option<String>,
    pub visibility: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ActivityView {
    pub id: i64,
    pub team_id: i64,
    pub creator_id: i64,
    pub title: String,
    pub location: LngLat,
    pub location_name: Option<String>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub content: Option<String>,
    pub notice: Option<String>,
    pub visibility: String,
    pub created_at: Option<DateTime<Utc>>,
}

impl From<ActivityRow> for ActivityView {
    fn from(r: ActivityRow) -> Self {
        Self {
            id: r.id,
            team_id: r.team_id,
            creator_id: r.creator_id,
            title: r.title,
            location: LngLat {
                lng: r.lng,
                lat: r.lat,
            },
            location_name: r.location_name,
            start_time: r.start_time,
            end_time: r.end_time,
            content: r.content,
            notice: r.notice,
            visibility: r.visibility,
            created_at: r.created_at,
        }
    }
}
